058-zk-kraft-migration.md

```diff
@@ -0,0 +1,437 @@
+
+In order to make the process clearer as much as possible, each phase is described by:
+
+* _Preconditions_: conditions that should be met (i.e. current state in the migration FSM) in order to have the operator taking actions on a trigger.
+* _Trigger_: what makes the operator actions to run only with verified preconditions.
+* _Actions_: what the operator does when preconditions are met (i.e. nodes configuration, rolling, ...) together with things that could be observable during actions (i.e. Kafka metrics changes).
+* _Status changes_: how the FSM changes status when the actions are completed.
+
+#### Already running KRaft-based clusters
+
+**_Preconditions_**
+
+The Apache Kafka cluster is running in KRaft mode, so the `strimzi.io/kraft: enabled` annotation is already applied on the `Kafka` custom resource.
+The FSM is in the `KRaft` state (as per `Kafka.status.metadataState` field).
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: migration` annotation on the `Kafka` custom resource.
+
+**_Actions_**
+
+The operator starts a reconciliation and updates the `Kafka.status` by adding a warning condition about the cluster being already KRaft-based, so no migration is possible.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM stays in the `KRaft` state.
+The `Kafka.status` keeps the warning condition, on each subsequent reconciliation, until the user re-applies the `strimzi.io/kraft: enabled` annotation.
+
+#### Phase 1 and 2: Provisioning the KRaft controller quorum, enabling the migration on the brokers or Rollback
+
+At the beginning of this phase, we expect the user to start the pre-migration for the current ZooKeeper-based cluster.
+
+**_Preconditions_**
+
+The Apache Kafka cluster is running in ZooKeeper mode, so the `strimzi.io/kraft: disabled` annotation is already applied on the `Kafka` custom resource.
+The FSM is in the `ZooKeeper` state (as per `Kafka.status.metadataState` field) and it keeps staying in this state until the user applies the `strimzi.io/kraft: migration` annotation on the `Kafka` custom resource, and creates the `KafkaNodePool` custom resources(s) for the KRaft controllers.
+
+**_Trigger_**
+
+The user creates the `KafkaNodePool` custom resource(s) containing the definition of the KRaft controllers which are going to replace the ZooKeeper nodes.
+The user sets the `strimzi.io/kraft: migration` annotation on the `Kafka` custom resource. 
+
+**_Actions_**
+
+The operator starts a reconciliation and, via the `KafkaReconciler` component, creates and configures the controllers accordingly with the ZooKeeper migration flag and the connection to the ZooKeeper ensemble.
+The controllers are automatically created with the following configuration (the user cannot set them manually).
+
+```shell
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
```

Comment:
Does Strimzi support using TLS/SASL when connecting to ZooKeeper?

Reply:
Yeah, all the `zookeeper.*` properties are set properly for TLS.